src/redisHashKeyValueDB.ts

```diff
@@ -0,0 +1,159 @@
+import {
+  CommonKeyValueDBSaveBatchOptions,
+  CommonDBCreateOptions,
+  CommonKeyValueDB,
+  KeyValueDBTuple,
+} from '@naturalcycles/db-lib'
+import { _chunk, _mapObject, StringMap } from '@naturalcycles/js-lib'
+import { ReadableTyped } from '@naturalcycles/nodejs-lib'
+import { RedisClient } from './redisClient'
+import { RedisKeyValueDBCfg } from './redisKeyValueDB'
+
+export interface RedisHashKeyValueDBCfg extends RedisKeyValueDBCfg {
+  hashKey: string
+}
+
+export class RedisHashKeyValueDB implements CommonKeyValueDB, AsyncDisposable {
+  client: RedisClient
+  keyOfHashField: string
+
+  constructor(cfg: RedisHashKeyValueDBCfg) {
+    this.client = cfg.client
+    this.keyOfHashField = cfg.hashKey
+  }
+
+  async ping(): Promise<void> {
+    await this.client.ping()
+  }
+
+  async [Symbol.asyncDispose](): Promise<void> {
+    await this.client.disconnect()
+  }
+
+  async getByIds(table: string, ids: string[]): Promise<KeyValueDBTuple[]> {
+    if (!ids.length) return []
+    // we assume that the order of returned values is the same as order of input ids
+    const bufs = await this.client.hmgetBuffer(this.keyOfHashField, this.idsToKeys(table, ids))
+    return bufs.map((buf, i) => [ids[i], buf] as KeyValueDBTuple).filter(([_k, v]) => v !== null)
+  }
+
+  async deleteByIds(table: string, ids: string[]): Promise<void> {
+    if (!ids.length) return
+    await this.client.hdel(this.keyOfHashField, this.idsToKeys(table, ids))
+  }
+
+  async saveBatch(
+    table: string,
+    entries: KeyValueDBTuple[],
+    opt?: CommonKeyValueDBSaveBatchOptions,
+  ): Promise<void> {
+    if (!entries.length) return
+
+    const map: StringMap<any> = _mapObject(Object.fromEntries(entries), (k, v) => [
+      this.idToKey(table, String(k)),
+      v,
+    ])
+
+    if (opt?.expireAt) {
+      await this.client.hsetWithTTL(this.keyOfHashField, map, opt.expireAt)
+    } else {
+      await this.client.hset(this.keyOfHashField, map)
+    }
+  }
+
+  streamIds(table: string, limit?: number): ReadableTyped<string> {
+    let stream = this.client
+      .hscanStream(this.keyOfHashField, {
+        match: `${table}:*`,
+      })
+      .flatMap(keyValueList => {
+        const keys: string[] = []
+        keyValueList.forEach((keyOrValue, index) => {
+          if (index % 2 !== 0) return
+          keys.push(keyOrValue)
+        })
+        return this.keysToIds(table, keys)
+      })
+
+    if (limit) {
+      stream = stream.take(limit)
+    }
+
+    return stream
+  }
+
+  streamValues(table: string, limit?: number): ReadableTyped<Buffer> {
+    return this.client
+      .hscanStream(this.keyOfHashField, {
+        match: `${table}:*`,
+      })
+      .flatMap(
+        async keyValueList => {
```

Comment:
Q: why is it async (and has concurrency), while I see no await inside?

Reply:
Copy-pasted from `redisKeyValueDB` where it is correctly async.
As I re-wrote the function, I didn't notice that I removed the awaited calls in it.